### PR TITLE
Implement routes for refresh token authentication

### PR DESCRIPTION
**Pull request by lawliet89**

### CODE REVIEW

File: src/cors.rs [hunk 1]
```diff
@@ -337,19 +337,16 @@ impl AllowedOrigins {
     /// Returns a tuple where the first element is the struct `AllowedOrigins`, and the second element
     /// is a map of strings which failed to parse into URLs and their associated parse errors.
     pub fn new_from_str_list(urls: &[&str]) -> (Self, HashMap<String, ParseError>) {
-        let urls: HashMap<String, Result<Url, ParseError>> = urls.iter()
```
Comment (lawliet89) [inline src/cors.rs:340]: This is just a tiny refactor based on what I spotted.

File: src/token.rs [hunk 2]
```diff
@@ -775,26 +904,68 @@ impl Default for Secret {
...
+        }
+    }
+
+    /// Create a JWK for the purpose of decryption
+    pub(super) fn for_decryption(&self) -> Result<jwk::JWK<jwt::Empty>, Error> {
+        // For now
+        self.for_encryption()
+    }
+
+    pub(super) fn read_file_to_bytes(path: &str) -> Result<Vec<u8>, Error> {
```
Comment (lawliet89) [inline src/token.rs]: Should be private

File: src/token.rs [hunk 3]
```diff
@@ -396,12 +485,55 @@ impl RefreshToken {
...
+            .as_ref()
+            .ok_or_else(|| Error::InvalidAudience)?;
+
+        verify_service(config, service)
+            .and_then(|_| if audience.contains(&FromStr::from_str(service)?) {
+                          Ok(())
+                      } else {
+                          Err(Error::InvalidAudience)
+                      })
+            .and(verify_audience(config, audience))
```
Comment (lawliet89) [inline src/token.rs]: Should use and then for lazy evaluation

File: src/routes.rs [hunk 4]
```diff
@@ -112,11 +116,40 @@ fn refresh_token(origin: Option<cors::Origin>,
...
                  -> Result<cors::Response<Token<PrivateClaim>>, ::Error> {
 
+    if !configuration.refresh_token_enabled() {
+        return Err(::Error::BadRequest("Refresh token is not enabled".to_string()));
+    }
+    let refresh_token_configuration = configuration.refresh_token();
+
     auth_param.verify(&authorization)?;
-    Err(::Error::UnsupportedOperation)
+    let refresh_token = RefreshToken::new_encrypted(&authorization.token());
```
Comment (gyng) [inline src/routes.rs:129]: Why encrypt then decrypt?

File: src/token.rs [hunk 5]
```diff
@@ -153,6 +172,36 @@ fn make_token<P: Serialize + Deserialize + 'static>(subject: &str,
                              }))
 }
 
+/// Verify that the service requested for is allowed in the configuration
+fn verify_service(config: &Configuration, service: &str) -> Result<(), Error> {
+    if !config.audience.contains(&FromStr::from_str(service)?) {
```
Comment (gyng) [inline src/token.rs:177]: Is this correct? If my audience is `http://www.example.com` and my service is `http://www.e` it returns `Ok(())`?

**Review by gyng — APPROVED** [review]

File: src/routes.rs (hunk 4 above)
Comment (lawliet89) [reply src/routes.rs:129]: The token was received in an encrypted form. This is calling the enum's "constuctor" to create the encrypted variant. See https://github.com/lawliet89/rowdy/pull/23/files#diff-8eaf0fafad7d2ed9ec9a701c04424c96R412

File: src/token.rs (hunk 5 above)
Comment (lawliet89) [reply src/token.rs:177]: Yes. This is not `String::contains`. It's this: https://lawliet89.github.io/biscuit/biscuit/enum.SingleOrMultiple.html#method.contains

Comment (lawliet89) [reply src/token.rs:177]: Playground: https://play.rust-lang.org/?gist=21965509985b3cb6aeefb857c4604d53&version=stable&backtrace=0

**Pull request merged**